contracts/leaser/src/migrate.rs

```diff
@@ -2,101 +2,305 @@ use lease::api::MigrateMsg;
 use platform::batch::Batch;
 use sdk::cosmwasm_std::Addr;
 
-use crate::{error::ContractError, result::ContractResult};
+use crate::{msg::NbInstances, result::ContractResult};
 
-pub fn migrate_leases<I>(leases: I, lease_code_id: u64) -> ContractResult<Batch>
+pub struct Customer<LeaseIter> {
+    customer: Addr,
+    leases: LeaseIter,
+}
+
+#[derive(Default)]
+#[cfg_attr(test, derive(Debug, Eq, PartialEq))]
+pub struct MigrationResult {
+    pub msgs: Batch,
+    pub next_customer: Option<Addr>,
+}
+
+/// Builds a batch of messages for the migration of up to `max_leases`
+///
+/// Consumes the customers iterator to the next customer or error.
+pub fn migrate_leases<I, LI>(
+    mut customers: I,
+    lease_code_id: u64,
+    max_leases: NbInstances,
+) -> ContractResult<MigrationResult>
 where
-    I: Iterator<Item = ContractResult<Addr>>,
+    I: Iterator<Item = ContractResult<Customer<LI>>>,
+    LI: ExactSizeIterator<Item = Addr>,
 {
-    let no_msgs = MigrateBatch::new(lease_code_id);
-    let migrated_msgs = leases.fold(no_msgs, MigrateBatch::migrate_lease);
-    migrated_msgs.try_into()
+    let mut msgs = MigrateBatch::new(lease_code_id, max_leases);
+    customers
+        .find_map(|may_customer| match may_customer {
```

Comment:
(Typo) Sadly, May the 4th has passed us already. ;)
```suggestion
        .find_map(|maybe_customer| match maybe_customer {
```